File: src/robot_control_pkg/ur5_rtde_control/ur5_rtde_control/test_rtde_performance.py
import time
import numpy as np
from ur_rtde import RTDEControl, RTDEReceive

ROBOT_IP = "192.168.1.102"  # 修改为你的UR机器人IP
FREQUENCIES = [10, 50, 100, 125, 200, 250, 500, 750, 1000]  # Hz
TEST_DURATION = 10  # 每个频率测试时长（秒）


def test_rtde_performance():
    print("连接到UR机器人...", ROBOT_IP)
    rtde_c = RTDEControl(ROBOT_IP)
    rtde_r = RTDEReceive(ROBOT_IP)
    print("连接成功！")

    results = []
    for freq in FREQUENCIES:
        print(f"\n测试频率: {freq} Hz")
        period = 1.0 / freq
        timestamps = []
        delays = []
        lost = 0
        start_time = time.time()
        count = 0
        while time.time() - start_time < TEST_DURATION:
            t0 = time.time()
            try:
                # 这里只做简单的getActualQ()，可换成moveJ等命令
                _ = rtde_r.getActualQ()
                t1 = time.time()
                delay = t1 - t0
                delays.append(delay)
                timestamps.append(t0)
            except Exception as e:
                print(f"通信异常: {e}")
                lost += 1
            count += 1
            elapsed = time.time() - t0
            if elapsed < period:
                time.sleep(period - elapsed)
        if delays:
            delays_np = np.array(delays)
            result = {
                "freq": freq,
                "count": count,
                "lost": lost,
                "max_delay": np.max(delays_np),
                "min_delay": np.min(delays_np),
                "avg_delay": np.mean(delays_np),
                "std_delay": np.std(delays_np),
            }
        else:
            result = {"freq": freq, "count": count, "lost": lost, "max_delay": None, "min_delay": None, "avg_delay": None, "std_delay": None}
        results.append(result)
        print(f"频率: {freq}Hz, 总包数: {count}, 丢包: {lost}, 最大延时: {result['max_delay']:.6f}s, 最小延时: {result['min_delay']:.6f}s, 平均延时: {result['avg_delay']:.6f}s")

    print("\n=== 测试结果汇总 ===")
    for r in results:
        print(r)

    rtde_c.disconnect()
    rtde_r.disconnect()

if __name__ == "__main__":
    test_rtde_performance()

File: src/robot_control_pkg/ur5_rtde_cpp/src/test_rtde_performance.cpp
// C++ RTDE通信性能测试程序
// 依赖：ur_rtde C++接口（需提前准备好），可用chrono计时
#include <iostream>
#include <vector>
#include <chrono>
#include <thread>
#include <numeric>
#include <algorithm>
// #include <ur_rtde/rtde_control_interface.h>
// #include <ur_rtde/rtde_receive_interface.h>

int main(int argc, char **argv) {
    // 参数
    std::string robot_ip = "192.168.1.102";
    std::vector<int> frequencies = {10, 50, 100, 125, 200, 250, 500, 750, 1000};
    int test_duration = 10; // 秒

    // std::cout << "连接到UR机器人..." << robot_ip << std::endl;
    // ur_rtde::RTDEControlInterface rtde_c(robot_ip);
    // ur_rtde::RTDEReceiveInterface rtde_r(robot_ip);
    // std::cout << "连接成功！" << std::endl;

    for (int freq : frequencies) {
        std::cout << "\n测试频率: " << freq << " Hz" << std::endl;
        double period = 1.0 / freq;
        std::vector<double> delays;
        int lost = 0;
        int count = 0;
        auto start_time = std::chrono::steady_clock::now();
        while (std::chrono::duration_cast<std::chrono::seconds>(std::chrono::steady_clock::now() - start_time).count() < test_duration) {
            auto t0 = std::chrono::steady_clock::now();
            try {
                // auto q = rtde_r.getActualQ(); // 需替换为C++接口
                std::this_thread::sleep_for(std::chrono::milliseconds(1)); // 模拟调用
                auto t1 = std::chrono::steady_clock::now();
                double delay = std::chrono::duration<double>(t1 - t0).count();
                delays.push_back(delay);
            } catch (...) {
                std::cout << "通信异常" << std::endl;
                lost++;
            }
            count++;
            double elapsed = std::chrono::duration<double>(std::chrono::steady_clock::now() - t0).count();
            if (elapsed < period) {
                std::this_thread::sleep_for(std::chrono::duration<double>(period - elapsed));
            }
        }
        if (!delays.empty()) {
            double max_delay = *std::max_element(delays.begin(), delays.end());
            double min_delay = *std::min_element(delays.begin(), delays.end());
            double avg_delay = std::accumulate(delays.begin(), delays.end(), 0.0) / delays.size();
            double std_delay = 0.0;
            for (auto d : delays) std_delay += (d - avg_delay) * (d - avg_delay);
            std_delay = std::sqrt(std_delay / delays.size());
            std::cout << "频率: " << freq << "Hz, 总包数: " << count << ", 丢包: " << lost
                      << ", 最大延时: " << max_delay << "s, 最小延时: " << min_delay
                      << "s, 平均延时: " << avg_delay << "s" << std::endl;
        }
    }
    // rtde_c.disconnect();
    // rtde_r.disconnect();
    return 0;
}
